sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,662 @@
+		// between `receive_frame` call and we marking interrupt as received,
+		// it will trigger again immediately after we acknowledge it.
+
+		// Acknowledge the interrupt in the scheduler.
+		interrupt_complete(STATIC_SEALED_VALUE(EthernetInterruptCap));
+		if (*receiveInterruptFutex == lastInterruptValue)
+		{
+			Debug::log("Acknowledged interrupt, sleeping on futex {}",
+			           receiveInterruptFutex);
+			return futex_timed_wait(
+			  timeout, receiveInterruptFutex, lastInterruptValue);
+		}
+		Debug::log("Scheduler announces interrupt has fired");
+		return 0;
+	}
+
+	/**
+	 * Simple class representing a received Ethernet frame.
+	 */
+	struct Frame
+	{
+		uint16_t length;
+		uint8_t *buffer;
+
+		Frame(uint16_t length) : length(length)
+		{
+			buffer = new (std::nothrow) uint8_t[length];
+		}
+
+		Frame(const Frame &) = delete;
+		Frame(Frame &&other) : buffer(other.buffer), length(other.length)
+		{
+			other.buffer = nullptr;
+			other.length = 0;
+		}
+
+		~Frame()
+		{
+			if (buffer != nullptr)
+			{
+				delete[] buffer;
+			}
+		}
+	};
+
+	/**
+	 * Check the link status of the PHY.
+	 */
+	bool phy_link_status()
+	{
+		uint16_t status = register_read(RegisterOffset::P1SR);
+		return (status & Port1Status::LinkGood) != 0;
+	}
+
+	std::optional<Frame> receive_frame()
+	{
+		if (frames_to_process == 0)
+		{
+			uint16_t isr = register_read(RegisterOffset::ISR);
+			if (!(isr & ReceiveInterrupt))
+			{
+				return std::nullopt;
+			}
+
+			// Acknowledge the interrupt
+			register_write(RegisterOffset::ISR, ReceiveInterrupt);
+
+			// Read number of frames pending.
+			// Note that this is only updated when we acknowledge the interrupt.
+			frames_to_process = register_read(RegisterOffset::RXFCTR) >> 8;
+		}
+
+		// Get number of frames pending
+		for (; frames_to_process; frames_to_process--)
+		{
+			uint16_t status = register_read(RegisterOffset::RXFHSR);
+			uint16_t length = register_read(RegisterOffset::RXFHBCR) & 0xFFF;
+			bool     valid =
+			  (status & RxFrameValid) &&
+			  !(status &
+			    (RxCrcError | RxRuntFrame | RxFrameTooLong | RxMiiError |
+			     RxUdpFrameChecksumStatus | RxTcpFrameChecksumStatus |
+			     RxIpFrameChecksumStatus | RxIcmpFrameChecksumStatus));
+
+			if (!valid && DebugDroppedFrames)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with status: {}", status);
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			if (length == 0)
+			{
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame with zero length");
+				}
+
+				drop_error_frame();
+				continue;
+			}
+
+			Debug::log("Receiving frame of length {}", length);
+
+			auto frame = Frame(length);
+			if (!frame.buffer)
+			{
+				// Out of memory while trying to receive a packet.
+				// Drop the frame when this happens.
+				if (DebugDroppedFrames)
+				{
+					Debug::log("Dropping frame due to OOM");
+				}
+				drop_error_frame();
+				continue;
+			}
```

Comment:
If this is the contract between the firewall and the Ethernet driver, then the API design of `receive_frame` should be taking a callback instead of returning something of unconstrained lifetime.